src/sqlfluff/rules/L069.py

```diff
@@ -0,0 +1,175 @@
+        FROM foo;
+
+    """
+
+    groups = ("all",)
+    config_keywords = ["allow_non_standard_type_casting"]
+    crawl_behaviour = SegmentSeekerCrawler({"function", "cast_expression"})
+
+    def _eval(self, context: RuleContext) -> Optional[LintResult]:
+        """Use ``CAST`` instead of ``CONVERT`` or ``::``."""
+        # Config type hints
+        self.allow_non_standard_type_casting: bool
+
+        # Rule disabled for teradata.
+        if context.dialect.name == "teradata":
+            return None
+
+        if not self.allow_non_standard_type_casting:
+            # Limit scope to CONVERT and casting_operator(::)
+            # CONVERT
+            if (
+                context.segment.is_type("function")
+                and context.segment.get_child("function_name").raw_upper == "CONVERT"
+            ):
+                # Get the content of CONVERT
+                convert_content = (
+                    FunctionalContext(context)
+                    .segment.children(sp.is_type("bracketed"))
+                    .children(
+                        sp.and_(
+                            sp.not_(sp.is_meta()),
+                            sp.not_(
+                                sp.is_type(
+                                    "start_bracket",
+                                    "end_bracket",
+                                    "whitespace",
+                                    "newline",
+                                    "comma",
+                                )
+                            ),
+                        )
+                    )
+                )
+
+                # we only care about 2-arguments convert
+                # some dialects allow an optional 3rd argument e.g TSQL
+                # which cannot be rewritten into CAST
+                if len(convert_content) > 2:
+                    return None
+
+                # get the expression
+                expression = convert_content[1].raw
+
+                # get the data type
+                datatype = convert_content[0].raw
+
+                # Create fix to replace ``CONVERT`` with ``CAST``.
+                fix = LintFix.replace(
+                    context.segment,
+                    [
+                        CodeSegment(
+                            raw=f"CAST({expression} AS {datatype})",
+                        )
+                    ],
+                )
+
+                return LintResult(
+                    anchor=context.segment,
+                    fixes=[fix],
+                    description=f"Use 'CAST' instead of \n"
+                    f"'{context.segment.get_child('function_name').raw_upper}'.",
+                )
+
+            # casting_operator(::)
+            if context.segment.is_type("cast_expression"):
+                # get the expression and the datatype segment
+                expression_datatype_segment = FunctionalContext(
+                    context
+                ).segment.children(
```

Comment:
Nitpick: This code is very similar to lines 79-94. Possible to avoid the duplication?

Reply:
Yes, I agree. The main issue is the bracketed content in `convert`. I don't see an easy way around it.

Reply:
You could create a separate function with the code inside it, e.g.:
```
def do_something(segments: Segments) -> Segments:
return segments.children(                    sp.and_(
                        sp.not_(sp.is_meta()),
                        sp.not_(
                            sp.is_type(
                                "start_bracket",
                                "end_bracket",
                                "whitespace",
                                "newline",
                                "casting_operator",
                            )
                        ),
                    )
                )
```

then call that function from both places.

Reply:
Fixed by [ebeb4ad](https://github.com/sqlfluff/sqlfluff/pull/3747/commits/ebeb4ada735810c3d050d690d93e0d8fcc64d4fa)